pgconn/pgconn.go

```diff
@@ -1740,6 +1741,30 @@ func (pgConn *PgConn) flushWithPotentialWriteReadDeadlock() error {
 	return err
 }
 
+// SyncConn prepares the underlying net.Conn for direct use. PgConn may internally buffer reads or use goroutines for
+// background IO. This means that any direct use of the underlying net.Conn may be corrupted if a read is already
+// buffered or a read is in progress. SyncConn drains read buffers and stops background IO. In some cases this may
+// require sending a ping to the server. ctx can be used to cancel this operation. This should be called before any
+// operation that will use the underlying net.Conn directly. e.g. Before Conn() or Hijack().
+//
+// This should not be confused with the PostgreSQL protocol Sync message.
+func (pgConn *PgConn) SyncConn(ctx context.Context) error {
```

Comment:
I am not sure if the description is correct here - the background reader continues as long as there was no error, or the state is still running, so without calling `Close()` first, how would the background reader actually close ?

This relates to the test as well - since you never call close on the background reader, `SyncConn` will never finish since there is no signal to the background reader to stop after one read.

Reply:
🤦‍♂️ I missed stopping the background reader. It was never supposed to run continuously it was supposed to stop as soon as possible after `exitPostentialWriteReadDeadlock` is called (i.e. after the next read).

That's what I was relying on with the design of `SyncConn`. If `bgReader` is running then a `Ping` will trigger a read which will terminate the background read. It loops several times in case a LISTEN/NOTIFY or other server initiated message is received.

I think it should work now with c8656db84c492e4400f72aa6cf240d373e4822d1.